### PR TITLE
Change logic for routing withdraw requests

### LINKED ISSUE #14129: Hearing withdraw routing logic
## Description

For Legacy appeals with powers of attorney (POAs) that are private attorneys/agents and state/regional organizations, route the appeal to VACOLS location 81 (Case Storage) after the hearing request is withdrawn.

Currently cases with private attorney POAs are sent to VACOLS location 55 when they should have been sent to 81, which causes the case to be sent to a queue for veteran service organization (VSO) to write an informal hearing proposal (IHP) for it. If the Veteran isn't represented by a VSO and they are sent to location 55 in VACOLS, the record sits there for a long time until someone pulls a "55 report" and sees that it's been sitting for X days.

This misrouting requires Hearing Coordinators to go through a long flow in VACOLS to reroute the cases and encourages some to not use the Caseflow tools available.

If the appeal is Legacy, assign it to:
- 55 if the POA / representative type is a Co-located Veteran Service Organization (VSO)
- 81 (Case Storage) if the POA / representative type is any other, including:
  - private agent / attorney
  - state or regional VA organization (e.g. NC Department of Military & Veterans Affairs)
  - no representative
  - a non-co-located Veteran Service Organization (VSO)
- If the appeal is AMA, assign it to Mail (create an EvidenceSubmissionWindow task, which is part of the Mail org)

## Acceptance criteria
- [ ] Correct routing to location 81

## Background/context/resources
According to Caseflow documentation, this routing should already be happening: 
https://github.com/department-of-veterans-affairs/appeals-team/blob/master/Project%20Folders/Caseflow%20Projects/Hearings/Hearing%20Schedule/Tech%20Specs/HearingTaskFlow.md#task-hierarchy

Legacy Appeal Flow:
- **Withdrawn Hearing Request (No Hearing Scheduled) Flow** - The user cancels the ScheduleHearingTask, the Hearing task is automatically closed, and we move the appeal to location 81 (case storage) or 55 (if VSO or co-located)
- **Canceled Hearing Flow** - We close the DispositionTask, the HearingTask is automatically closed, and we move the appeal to location 81 (case storage) or 55 (if VSO or co-located)

AMA Appeal Flow:
- **Withdrawn Hearing Request (No Hearing Scheduled) Flow** - The user cancels the ScheduleHearingTask and the HearingTask is automatically closed"
- **Canceled Hearing Flow** - We cancel the DispositionTask and the HearingTask is automatically closed

## Technical notes
- [POA hash](https://github.com/department-of-veterans-affairs/caseflow/blob/ce89d4c8a4d4bf3234c221518ffc665bfd62fe48/app/mappers/power_of_attorney_mapper.rb#L150-L156)
- Line 97 in the ScheduleHearingTask model
- Possibility to reverse logic

## Open questions
- [ ] Can we tell the difference between VSO and state/regional VA or state or regional VA organization?
- BGS has the distinction but when pulling it into Caseflow, we don't have that level of granularity.

## Related tickets (routing)
- Introduce granularity between VSOs and state / regional VA organization #14297
- Allow Hearing Branch to change docket without cancelling or postponing #12653
- Automatically transfer No Show hearing tasks #12539

### LINKED ISSUE #14297: Introduce granularity between VSOs and state / regional VA organization
## Description
Introduce granularity in Caseflow between _Co-located Veteran Service Organizations (VSOs)_ and _State or Regional VA Organizations_ in order to better route cases to different locations upon hearing withdraw, no-show, etc.

This work should allow Caseflow to perform the following actions, and will allow the work described in #14129 and #12539 to be completed.

If the appeal is Legacy and being withdrawn or canceled, assign it to:
- 55 if the POA / representative type is a Co-located Veteran Service Organization (VSO)
- 81 (Case Storage) if the POA / representative type is any other, including:
  - private agent / attorney
  - state or regional VA organization (e.g. NC Department of Military & Veterans Affairs)
  - no representative
  - a non-co-located Veteran Service Organization (VSO)
- If the appeal is AMA, assign it to Mail (create an EvidenceSubmissionWindow task, which is part of the Mail org)

## Acceptance criteria
- [ ] Update [Hearing Task Flow documentation](https://github.com/department-of-veterans-affairs/appeals-team/blob/master/Project%20Folders/Caseflow%20Projects/Hearings/Hearing%20Schedule/Tech%20Specs/HearingTaskFlow.md)

## Technical notes
TK

## Open questions
- Is there a better way for fetching information from BGS other than the roundabout way of getting info? 
- Once we have it, what do we do with it? Currently don't classify them

## Next steps
- define the research that needs to happen to pull co-located VSO organization list from VACOLS

### PR DESCRIPTION
**Pull request by sahalliburton**
Resolves #14129 
Resolves #14297 

### Description
Reverses the check of empty representatives to check whether the representative is part of a Vso and route to service organization location

### Acceptance Criteria
- [x] Code compiles correctly

### CODE REVIEW

**Review by tomas-nava — CHANGES_REQUESTED** [review]
I think this is looking good! I've requested a few changes.

Also, #14129 isn't fully resolved by this PR; I think we should say it Connects instead. Once we've completed #14297, we'll be able to resolve.

### the way it works before this PR is merged
```markdown
if there is no representative
then assign the appeal to location *81* (case storage)
else assign the appeal to location *55* (service organization)
```
**What's missing:** Appeals with private agent/attorney and state/regional VA organization representatives are being assigned to location 55 instead of 81.

### the way it will work after this PR merges
```markdown
if the POA/representative type is a VSO
then assign the appeal to location *55* (service organization)
else assign the appeal to location *81* (case storage)
```
**What's missing:** Appeals with non-colocated VSO representatives will be assigned to location 55 instead of 81.

### the way it must work before we can resolve
(after #14297)
```markdown
if the POA/representative type is a colocated VSO
then assign the appeal to location *55* (service organization)
else assign the appeal to location *81* (case storage)
```

### edit
This now resolves #14297 as well as #14129 as of 77884f31a783bc5d697af642b74ebaa017ec1ed6

**Comment by tomas-nava** [comment]
This is blocked because the location is being updated twice every time a hearing is withdrawn; once in `ScheduleHearingTask.withdraw_hearing` when the hearing is withdrawn, and then in `HearingTask.when_child_task_completed` after the `ScheduleHearingTask` is closed. The two locations may conflict (as they do in the test: `spec/models/tasks/schedule_hearing_task_spec.rb:202`)

**Review by tomas-nava — COMMENTED** [review]
removing the request changes requirement

**Review by tomas-nava — APPROVED** [review]
really removing the request changes requirement

File: spec/models/tasks/hearing_task_spec.rb [hunk 1]
```diff
@@ -69,4 +69,91 @@
...
+
+      before do
+        allow(LegacyAppealRepresentative).to receive(:new).and_return(lar)
+      end
+
+      context "hearing is held" do
+        let(:disposition) { "H" }
+
+        it "moves the appeal to transcription" do
+          expect { subject }.to change { vacols_case.reload.bfcurloc }.from(loc_caseflow).to(loc_transcription)
```
Comment (ferristseng) [inline spec/models/tasks/hearing_task_spec.rb:105]: Whoa, TIL `change...from...to`!

**Review by ferristseng — APPROVED** [review]
LGTM! Tests look great! :shipit:

**Comment by lund5000** [comment]
@tomas-nava We have approval from John Gosnell to release this to production.

**Pull request merged**